src/services/dexes/uniswapV3/UniswapV3MultiFetcher.ts

```diff
@@ -0,0 +1,148 @@
+import {StaticJsonRpcProvider} from '@ethersproject/providers';
+import {inject, injectable} from 'inversify';
+import {Logger} from 'winston';
+import {BigNumber} from 'ethers';
+import {readFileSync} from 'fs';
+import {fileURLToPath} from 'url';
+import path from 'path';
+
+import {DexProtocolName} from '../../../types/Dexes.js';
+import {ChainsIds} from '../../../types/ChainsIds.js';
+import {UniswapV3PoolRepository} from '../../../repositories/UniswapV3PoolRepository.js';
+import Settings from '../../../types/Settings.js';
+import {ContractAddressService} from '../../../services/ContractAddressService.js';
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+
+export type UniswapV3MultiFetcherParams = {
+  fromChain: string[];
+  base: string;
+  quote: string;
+  amountInDecimals: number;
+};
+
+type UniswapV3ContractHelperInput = {
+  pool: string;
+  base: string;
+  quote: string;
+  amountInDecimals: number;
+};
+export interface OutputValues {
+  base: string;
+  quote: string;
+  value: number;
+}
+
+@injectable()
+class UniswapV3MultiFetcher {
+  provider: StaticJsonRpcProvider | undefined;
+  readonly dexProtocol = DexProtocolName.UNISWAP_V3;
+
+  @inject('Logger') protected logger!: Logger;
+  @inject('Settings') protected settings!: Settings;
+  @inject(UniswapV3PoolRepository) protected uniswapV3PoolRepository!: UniswapV3PoolRepository;
+  @inject(ContractAddressService) contractAddressService!: ContractAddressService;
+
+  async apply(inputs: UniswapV3MultiFetcherParams[]): Promise<OutputValues[]> {
+    this.logger.debug(`[UniswapV3MultiFetcher]: start with inputs ${JSON.stringify(inputs)}`);
+
+    const poolsToFetch = await this.getPoolsToFetch(inputs);
+
+    if (poolsToFetch.size === 0) {
+      this.logger.error('[UniswapV3MultiFetcher] no data to fetch');
+      return [];
+    }
+
+    const prices = await Promise.all(
+      [...poolsToFetch.entries()].map(([chainId, pools]) => this.fetchData(chainId, pools)),
+    );
+
+    return prices.flat();
+  }
+
+  public async getPoolsToFetch(
+    inputs: UniswapV3MultiFetcherParams[],
+  ): Promise<Map<ChainsIds, UniswapV3ContractHelperInput[]>> {
+    const helperInputMap: Map<ChainsIds, UniswapV3ContractHelperInput[]> = new Map();
+    const liquidityFreshness = this.settings.api.liquidityFreshness;
+    const liquidityUpdatedLimit = new Date(Date.now() - liquidityFreshness);
+
+    for (const input of inputs) {
+      const {fromChain, base, quote, amountInDecimals} = input;
+
+      const data = await this.uniswapV3PoolRepository.findUpdatedLiquidity({
+        protocol: this.dexProtocol,
+        fromChain,
+        base,
+        quote,
+        liquidityUpdatedLimit,
+      });
+
+      if (!data) {
+        this.logger.error(`[UniswapV3MultiFetcher] no data found for ${base}-${quote}`);
+        continue;
+      }
+
+      this.logger.debug(`[UniswapV3MultiFetcher] data found ${base}-${quote}`);
```

Comment:
lets add what data, this log is not that usefull without it

Reply:
Done